### PR TITLE
Variable Viewport Port

### PR DESCRIPTION
**Pull request by nevimer**
## About The Pull Request

This is a port of the variable viewport from Nebula. I've done my own tweaks to the skin and modifications to the original code to adapt it here. A bit of a spring cleaning on the UI, too!

https://user-images.githubusercontent.com/77420409/157742857-f1286ddb-3f7a-4458-adaf-8dceadfd5600.mp4

## Why It's Good For The Game

It's a massive modernization to the games interface, it scales like you would expect a modern game/program/anything to. No-one sane is writing things for fixed screen resolutions anymore.

club-penguin-cleaning

## Changelog
:cl:
add: Viewport now scales with windows gui elements. The game now scales properly to any screen resolution and icon size combo. 

refactor: In game GUI positions & skin have been redone
config: New config options for the max view range. 
server: New config options for the max view range. Might be expensive on massive sizes?
/:cl:

### CODE REVIEW

File: code/_onclick/hud/hud.dm [hunk 1]
```diff
@@ -31,40 +31,46 @@ GLOBAL_DATUM_INIT(global_hud, /datum/global_hud, new)
 	var/obj/screen/orange
 	var/obj/screen/holomap
 
-/datum/global_hud/proc/setup_overlay(var/icon_state)
+/datum/global_hud/proc/setup_overlay(var/icon_state, color)
 	var/obj/screen/screen = new /obj/screen()
-	screen.screen_loc = "1,1"
+	screen.screen_loc = "CENTER-7,CENTER-7"
```
Comment (nevimer) [inline code/_onclick/hud/hud.dm:36]: From now on, you muse use CENTER-7, CENTER-7 instead of 1,1 for fullscreens.

File: code/_onclick/hud/hud.dm [hunk 2]
```diff
@@ -120,7 +122,7 @@ GLOBAL_DATUM_INIT(global_hud, /datum/global_hud, new)
 	O.plane = PLANE_FULLSCREEN
 
 	//welding mask overlay black/dither
-	darkMask = newlist(/obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen)
+/* 	darkMask = newlist(/obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen, /obj/screen) // We must fix this.
```
Comment (nevimer) [inline code/_onclick/hud/hud.dm:125]: scuffed

File: code/modules/client/viewport.dm [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		set category = "Debug"
+
+		var/divisor = text2num(winget(src, "mapwindow.map", "icon-size")) || world.icon_size
+		var/winsize_string = winget(src, "mapwindow.map", "size")
+
+		to_chat(usr, "Current client view: [view]")
+		to_chat(usr, "Icon size: [divisor]")
+		to_chat(usr, "xDim: [round(text2num(winsize_string) / divisor)]")
+		to_chat(usr, "yDim: [round(text2num(copytext(winsize_string,findtext(winsize_string,"x")+1,0)) / divisor)]")
+/* 	proc/set_skybox_offsets(var/x_dim, var/y_dim) // This is fucky.
```
Comment (nevimer) [inline code/modules/client/viewport.dm:73]: doesn't work right.

File: code/modules/clothing/spacesuits/rig/rig.dm [hunk 4]
```diff
@@ -1160,7 +1160,7 @@
 
 //Boot animation screen objects
 /obj/screen/rig_booting
-	screen_loc = "1,1"
```
Comment (nevimer) [inline code/modules/clothing/spacesuits/rig/rig.dm:1163]: death to 1,1

**Review by nevimer — COMMENTED** [review]
:)

**Comment by Zandario** [comment]
Please don't include the variable zoom-out, you can probably leave the code in for it to be used on items, but zooming out the entire screen should be relegated to abilities and items, Scaling on the X-Axis is what we mainly want here

**Comment by nevimer** [comment]
> Please don't include the variable zoom-out, you can probably leave the code in for it to be used on items, but zooming out the entire screen should be relegated to abilities and items, Scaling on the X-Axis is what we mainly want here

If I did that, it would break support for users with bigger screen resolutions. This adds a server side config option for max tiles visible in X and Y. 

You need to understand this changes how the viewport scales, it finds the boundaries and expands it. In a default 15 x 15 x64 icons scale, the game sits comfortably at 1080p, and this would expand the game to the ends of the view port however it is dragged. However, at 4k, you need to run 128 icon scale for the same ratio. If you ran 128 at 1080p, you get all zoomed in. and for smaller screens? running a lower icon scale can help you see just as much as your bigger screen friends. 


Use the config options to limit range in tiles each direction.

**Comment by nevimer** [comment]
By default the limit is byond max, but set it to like 25x21 or something.

**Comment by lolman360** [comment]
does fit-viewport still worki

**Comment by nevimer** [comment]
> does fit-viewport still worki

ye, you can use it, turns your view into absolute instead of fractional values. this would make your viewport always auto fit every size change you do.

**Comment by nevimer** [comment]
im finishing this today, too many merge conflicts and I'll be improving much of this anywho to warrant a fresh PR.

**Pull request closed without merging**